### PR TITLE
[Api] Receiving discount promotion

### PR DESCRIPTION
**Pull request by Tomanhez**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

### CODE REVIEW

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 1]
```diff
@@ -310,6 +311,48 @@ public function thisItemShouldHaveCode(array $item, string $variantCode): void
...
+
+        foreach ($items as $item) {
+            if ($item['productName'] === $product->getName()) {
+                Assert::same($pricing->getPrice() - $amount, $item['subtotal']);
+            }
+
+            return;
+        }
+
+        throw new \InvalidArgumentException('Expected product does not exist');
```
Comment (AdamKasp) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:332]: if you create private function `isDecreasedBy(ProductInterface $product, int amount = 0): bool` you can ommit this duplicated code.

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 2]
```diff
@@ -530,4 +551,22 @@ private function checkProductQuantity(
...
     }
+
+    private function compareItemSubtotal(ProductInterface $product, \Closure $price): void
+    {
+        /** @var ChannelPricingInterface $pricing */
+        $pricing = $product->getVariants()->get(0)->getChannelPricingForChannel($this->sharedStorage->get('channel'));
+
+        $items = $this->responseChecker->getValue($this->cartsClient->getLastResponse(), 'items');
+
+        foreach ($items as $item) {
```
Comment (arti0090) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:562]:
```suggestion
        $productName = $product->getName();
        foreach ($items as $item) {
        if ($item['productName'] === $productName()) {
```
in this `$product->getName()` we will call once but I assume there is not a lot of items in the $items.

**Review by arti0090 — APPROVED** [review]

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php (hunk 2 above)
Comment (Tomanhez) [reply src/Sylius/Behat/Context/Api/Shop/CartContext.php:562]: You wrong, look in this scenario: https://github.com/Sylius/Sylius/pull/12081/files#diff-bb3427020ccf457607272ab922033ec2e77d77d2226fcd3e6e9089647b77a7b8L8-L9

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 3]
```diff
@@ -311,6 +312,26 @@ public function thisItemShouldHaveCode(array $item, string $variantCode): void
...
+            return $channelPricing->getPrice() - $amount;
+        });
+    }
+
+    /**
+     * @Then product :product price should not be decreased
+     */
+    public function productPriceShouldNotBeDecreased(ProductInterface $product): void
+    {
+        $this->compareItemSubtotal($product, function (ChannelPricingInterface $channelPricing) {
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:330]: Missing return type I guess

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 4]
```diff
@@ -541,4 +562,22 @@ private function checkProductQuantity(
             }
         }
     }
+
+    private function compareItemSubtotal(ProductInterface $product, \Closure $price): void
+    {
+        /** @var ChannelPricingInterface $pricing */
+        $pricing = $product->getVariants()->get(0)->getChannelPricingForChannel($this->sharedStorage->get('channel'));
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:569]:
```suggestion
        $pricing = $product->getVariants()->first()->getChannelPricingForChannel($this->sharedStorage->get('channel'));
```

**Review by GSadee — APPROVED** [review]

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 5]
```diff
@@ -541,4 +562,22 @@ private function checkProductQuantity(
...
+
+        $items = $this->responseChecker->getValue($this->cartsClient->getLastResponse(), 'items');
+
+        foreach ($items as $item) {
+            if ($item['productName'] === $product->getName()) {
+                Assert::same($price($pricing), $item['subtotal']);
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:575]:
```suggestion
                Assert::same($price($pricing), $item['total']);
```

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 6]
```diff
@@ -311,6 +312,26 @@ public function thisItemShouldHaveCode(array $item, string $variantCode): void
         );
     }
 
+    /**
+     * @Then /^(product "[^"]+") price should be decreased by ("[^"]+")$/
+     */
+    public function itsPriceShouldBeDecreasedBy(ProductInterface $product, int $amount): void
+    {
+        $this->compareItemSubtotal($product, function (ChannelPricingInterface $channelPricing) use ($amount) {
+            return $channelPricing->getPrice() - $amount;
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:321]: Won't it be easier if we would just pass an amount of discount as a second argument and make the calculation from this anonymous function in the private method?

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thank you, Tomasz! :1st_place_medal:

**Pull request merged**